controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -320,22 +330,30 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) scaleUpControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, numMachines int) error {
-	var errs []error
+func (r *KubeadmControlPlaneReconciler) scaleUpControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) error {
+	// Every control plane replica must be Ready
+	if int(kcp.Status.UnavailableReplicas) > 0 {
```

Comment:
Good question.

Right now, the reconciler updates the status whenever it's done reconciling. Ordinarily, that will happen whenever a Machine is added or deleted.

However, if the update is skipped for whatever reason, then `scaleUpControlPlane` will return an error, at which point the reconciler will update the status, and the next time around the test will pass.

Reply:
That is what I thought, which makes me think that we are essentially trusting the status value when we use it here. Instead I think we should either:
- attempt to determine this from a source of truth rather than the status from a previous reconciliation
- update the status prior to taking actions during reconciliation in addition to the update at the end of reconciliation.

Reply:
Addressed this by using control plane healthchecks https://github.com/kubernetes-sigs/cluster-api/pull/2319